consensus/reactor.go

```diff
@@ -762,967 +688,701 @@ func (conR *Reactor) gossipVotesForHeight(
+// be the case, and we spawn all the relevant goroutine to broadcast messages to
+// the peer. During peer removal, we remove the peer for our set of peers and
+// signal to all spawned goroutines to gracefully exit in a non-blocking manner.
+func (r *Reactor) processPeerUpdate(peerUpdate p2p.PeerUpdate) {
+	r.Logger.Debug("received peer update", "peer", peerUpdate.NodeID, "status", peerUpdate.Status)
 
-	return tmjson.Marshal(ps)
-}
-
-// GetHeight returns an atomic snapshot of the PeerRoundState's height
-// used by the mempool to ensure peers are caught up before broadcasting new txs
-func (ps *PeerState) GetHeight() int64 {
-	ps.mtx.Lock()
-	defer ps.mtx.Unlock()
-	return ps.PRS.Height
-}
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
 
-// SetHasProposal sets the given proposal as known for the peer.
-func (ps *PeerState) SetHasProposal(proposal *types.Proposal) {
-	ps.mtx.Lock()
-	defer ps.mtx.Unlock()
+	switch peerUpdate.Status {
+	case p2p.PeerStatusUp:
+		// Do not allow starting new broadcasting goroutines after reactor shutdown
+		// has been initiated. This can happen after we've manually closed all
+		// peer goroutines and closed r.closeCh, but the router still sends in-flight
+		// peer updates.
+		if !r.IsRunning() {
+			return
+		}
 
-	if ps.PRS.Height != proposal.Height || ps.PRS.Round != proposal.Round {
-		return
-	}
+		var (
+			ps *PeerState
+			ok bool
+		)
 
-	if ps.PRS.Proposal {
-		return
-	}
+		ps, ok = r.peers[peerUpdate.NodeID]
+		if !ok {
+			ps = NewPeerState(r.Logger, peerUpdate.NodeID)
+			r.peers[peerUpdate.NodeID] = ps
+		}
 
-	ps.PRS.Proposal = true
+		if !ps.IsRunning() {
+			// Set the peer state's closer to signal to all spawned goroutines to exit
+			// when the peer is removed. We also set the running state to ensure we
+			// do not spawn multiple instances of the same goroutines and finally we
+			// set the waitgroup counter so we know when all goroutines have exited.
+			ps.broadcastWG.Add(3)
+			ps.SetRunning(true)
+
+			// start goroutines for this peer
+			go r.gossipDataRoutine(ps)
+			go r.gossipVotesRoutine(ps)
+			go r.queryMaj23Routine(ps)
+
+			// Send our state to the peer. If we're fast-syncing, broadcast a
+			// RoundStepMessage later upon SwitchToConsensus().
+			if !r.waitSync {
+				go r.sendNewRoundStepMessage(ps.peerID)
+			}
+		}
 
-	// ps.PRS.ProposalBlockParts is set due to NewValidBlockMessage
-	if ps.PRS.ProposalBlockParts != nil {
-		return
-	}
+	case p2p.PeerStatusDown:
+		ps, ok := r.peers[peerUpdate.NodeID]
+		if ok && ps.IsRunning() {
+			// signal to all spawned goroutines for the peer to gracefully exit
+			ps.closer.Close()
 
-	ps.PRS.ProposalBlockPartSetHeader = proposal.BlockID.PartSetHeader
-	ps.PRS.ProposalBlockParts = bits.NewBitArray(int(proposal.BlockID.PartSetHeader.Total))
-	ps.PRS.ProposalPOLRound = proposal.POLRound
-	ps.PRS.ProposalPOL = nil // Nil until ProposalPOLMessage received.
-}
+			go func() {
```

Comment:
because we do it in a goroutine, is it possible to get another PeerStatusDown for the same peer and end up blocking in a second goroutine on `ps.broadcastWG.Wait()`, therefore leaking 1 goroutine?

Reply:
> is it possible to get another PeerStatusDown for the same peer

Yes. But I don't think this is a concern, one or both or many goroutines will block on `ps.broadcastWG.Wait()`, but once that waitgroup hits zero, they should all complete and the remaining logic becomes idempotent.